File: array/quick-sort/python/improvement.py
def partition(array, low, high):
  mid = (low + high) // 2
  candidates = [array[low], array[mid], array[high]]
  median = sorted(candidates)[1]

  if median == array[low]:
    [array[low], array[high]] = [array[high], array[low]]
  elif median == array[mid]:
    [array[mid], array[high]] = [array[high], array[mid]]

  pivot = array[high]
  i = low - 1

  for j in range(low, high):
    if array[j] <= pivot:
      i += 1
      array[i], array[j] = array[j], array[i]

  array[i+1], array[high] = array[high], array[i+1]

  return i+1

def quicksort(array, low=0, high=None):
  if high is None:
    high = len(array) - 1

  if low < high:
    pivot_index = partition(array, low, high)
    quicksort(array, low, pivot_index-1)
    quicksort(array, pivot_index+1, high)

def sort(array):
  quicksort(array)
  return array

my_array = [64, 34, 25, 12, 22, 11, 90, 5]
print("Sorted array:", sort(my_array))

File: array/quick-sort/js/improvement.js
function partition(array, low, high) {
  let mid = Math.floor((low + high) / 2);
  let cantidates = [array[low], array[mid], array[high]];
  cantidates.sort((a, b) => a - b); // sort in ascending order
  let median = cantidates[1];

  if (median === array[low]) {
    [array[low], array[high]] = [array[high], array[low]];
  } else if (median === array[mid]) {
    [array[mid], array[high]] = [array[high], array[mid]];
  }

  let pivot = array[high];
  let i = low - 1;

  for (let j = low; j < high; j++) {
    if (array[j] <= pivot) {
      i++;
      [array[i], array[j]] = [array[j], array[i]];
    }
  }

  [array[i + 1], array[high]] = [array[high], array[i + 1]];
  return i + 1;
}

function quickSort(array, low = 0, high = null) {
  if (high === null) {
    high = array.length - 1;
  }

  if (low < high) {
    let pivotIndex = partition(array, low, high);
    quickSort(array, low, pivotIndex - 1);
    quickSort(array, pivotIndex + 1, high);
  }
}


function sort(array) {
  let low = 0, high = array.length - 1;
  quickSort(array, low, high);
  return array;
}

console.log(sort([64, 34, 25, 12, 22, 11, 90, 5])); // [5, 11, 12, 22, 25, 34, 64, 90]
